cranelift/module/src/module.rs

```diff
@@ -384,11 +422,236 @@ impl Display for ModuleExtName {
+    /// different version of Cranelift.
+    // Note: This must be the first field to ensure that Serde will deserialize it before
+    // attempting to deserialize other fields that are potentially changed between versions.
+    _version_marker: VersionMarker,
+
     names: HashMap<String, FuncOrDataId>,
     functions: PrimaryMap<FuncId, FunctionDeclaration>,
     data_objects: PrimaryMap<DataId, DataDeclaration>,
 }
 
+#[cfg(feature = "enable-serde")]
+mod serialize {
+    // This is manually implementing Serialize and Deserialize to avoid serializing the names field,
+    // which can be entirely reconstructed from the functions and data_objects fields, saving space.
+
+    use super::*;
+
+    use serde::de::{Deserialize, Deserializer, Error, MapAccess, SeqAccess, Unexpected, Visitor};
+    use serde::ser::{Serialize, SerializeStruct, Serializer};
+    use std::fmt;
+
+    fn get_names<E: Error>(
+        functions: &PrimaryMap<FuncId, FunctionDeclaration>,
+        data_objects: &PrimaryMap<DataId, DataDeclaration>,
+    ) -> Result<HashMap<String, FuncOrDataId>, E> {
+        let mut names = HashMap::new();
+        for (func_id, decl) in functions.iter() {
+            if let Some(name) = &decl.name {
+                let old = names.insert(name.clone(), FuncOrDataId::Func(func_id));
+                if old.is_some() {
+                    return Err(E::invalid_value(
+                        Unexpected::Other("duplicate name"),
+                        &"FunctionDeclaration's with no duplicate names",
+                    ));
+                }
+            }
+        }
+        for (data_id, decl) in data_objects.iter() {
+            if let Some(name) = &decl.name {
+                let old = names.insert(name.clone(), FuncOrDataId::Data(data_id));
+                if old.is_some() {
+                    return Err(E::invalid_value(
+                        Unexpected::Other("duplicate name"),
+                        &"DataDeclaration's with no duplicate names",
+                    ));
+                }
+            }
+        }
+        Ok(names)
+    }
+
+    impl Serialize for ModuleDeclarations {
+        fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+            let ModuleDeclarations {
+                _version_marker,
+                functions,
+                data_objects,
+                names: _,
+            } = self;
+
+            let mut state = s.serialize_struct("ModuleDeclarations", 4)?;
+            state.serialize_field("_version_marker", _version_marker)?;
+            state.serialize_field("functions", functions)?;
+            state.serialize_field("data_objects", data_objects)?;
+            state.end()
+        }
+    }
+
+    enum ModuleDeclarationsField {
+        VersionMarker,
+        Functions,
+        DataObjects,
+        Ignore,
+    }
+
+    struct ModuleDeclarationsFieldVisitor;
+
+    impl<'de> serde::de::Visitor<'de> for ModuleDeclarationsFieldVisitor {
```

Comment:
The Deserialize impl is based on serde-derive's macro expansion, but it has been cleaned up a lot and modified to run `get_names` instead of expecting a `names` field. This should be license technically, right?

Reply:
Any chance you can use one of the facilities provided by serde, e.g. `#[serde(serialize_with = ...)]` on the `names` field, instead?

Reply:
When applied to the names field that doesn't give access to the functions and data_objects fields to reconstruct the names field from from what I can tell.